README.md
# Ikuchi - Simple Magnetosphere Viewer #
**Contributors: [Chris Arridge](https://www.lancaster.ac.uk/physics/about-us/people/chris-arridge), Lancaster University, c.arridge@lancaster.ac.uk; [Josh Wiggs](https://www.lancaster.ac.uk/physics/about-us/people/josh-wiggs), Lancaster University, j.wiggs@lancaster.ac.uk**

[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.3451443.svg)](https://doi.org/10.5281/zenodo.3451443)

`Ikuchi` is an HTML/Javascript application which uses [Three.JS](https://threejs.org/) to render 3D views of solar system magnetospheres. It is designed for quickly viewing the realistic 3D and time-dependent geometry of giant planet magnetospheres, particularly Uranus and Neptune. The scenes are viewed in a coordinate system with one axis pointing from the planet towards the Sun, and another pointing from the planet in the direction of its orbital motion. You may know this as a "Planetary Solar Orbital" coordinate system, e.g., Geocentric Solar Ecliptic (GSE) at Earth, or Kronocentric Solar Orbital (KSO) at Saturn. The orientation of the planets as a function of time is computed from non-linear functions fitted to data from NASA/JPL [Navigation and Ancillary Information Facility (NAIF)](https://naif.jpl.nasa.gov/naif/)'s [SPICE](https://naif.jpl.nasa.gov/naif/spiceconcept.html) system. The RMS errors associated with fitting to SPICE data are no larger than 0.06 deg, and most planets have a maximum errors no larger than around 0.03 degrees, apart from Jupiter and Saturn, which have maximum errors of around 0.2 degrees. Although the fits are generally very accurate, `Ikuchi` is not designed for accurate ephemeris. The orbits of natural satellites are especially simplified.

## Origins of the name `Ikuchi`
Ikuchi refers to a [supernatural sea creature](https://en.wikipedia.org/wiki/Ikuchi) from [Japanese legend](https://en.wikipedia.org/wiki/Y%C5%8Dkai). In some depictions it appears as a sort of Kraken with tentacles that it uses to envelop passing ships. The tentacles here reminded one of us of field lines in a very asymmetrical magnetosphere, hence the name of the tool.

![alt text](artwork.png "Artwork illustrating a planetary magnetosphere and sea-scene with one field line extending into a tentacle that plunges in and out of the waves of the sea.")

## Installation
Just copy the `Ikuchi` `source` folder somewhere on your machine. The folder can be renamed. The `Ikuchi.html` file must be in the same location as the `js` and `css` folders.

## Licence
Copyright (C) 2019-2023 Chris Arridge, Josh Wiggs and Lancaster University

This program is free software: you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free Software
Foundation, either version 3 of the License, or (at your option) any later version.

This program is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
GNU General Public License for more details.

You should have received a copy of the GNU General Public License
along with this program.  If not, see <https://www.gnu.org/licenses/>.

## Other credits
Texture maps for Earth, Jupiter, Saturn, Uranus and Neptune come from [Planet Pixel Emporium](http://planetpixelemporium.com/planets.html) and are copyright [James Hastings-Trew](https://twitter.com/jamesht?lang=en).

The Ikuchi artwork above uses a texture by [efe_madrid](https://www.freepik.com/efe-madrid) that the author kindly made freely available on [Freepik](https://www.freepik.com) The artwork itself was inspired by the works of <a href="https://en.wikipedia.org/wiki/Toriyama_Sekien">Sano Toyofusa</a> and <a href="https://en.wikipedia.org/wiki/Hokusai">Katsushika Hokusai</a>.</p>

## Usage
### Basic
By default, `Ikuchi` loads with a view of the dusk terminator showing a pseudo noon-midnight meridian view. The planet has zero obliquity (no seasons), a 10 degree dipole tilt, and a rotation period of 24 hours:

![alt text](default.png "Default view for Ikuchi")

The planet is rendered with a latitude-longitude grid, with the prime meridian and equator highlighted. The yellow lines are field lines. Four vectors show:
* Orange: a vector from the centre of the planet to the Sun.
* Crimson: the rotation axis of the planet.
* Blue: the dipole axis.
* Green: The orbital velocity vector of the planet.

Hence, the orange and green vectors identify the orbital plane of the planet, which is shown as a wire mesh by default. The equatorial plane of the planet is not visible in the default view, although this can be changed in the settings. The mouse can be used to manipulate the scene:
* Left-clicking and moving the mouse-pointer shifts the observer's view.
* Right-clicking and moving the mouse-pointer shifts the observer's location.
* The mouse wheel zooms in and out.

### GUI Controls
The GUI at the top right offers a range of options. For example, instead of using the mouse to move around the planet, the `Views` tab contains five pre-set views. The options panel also allows the user to swap between `perspective` and `orthographic` views. If you want to view the system in an isometric projection, change the camera to `orthographic` and select the `oblique` view.

By default the planet auto-rotates at the specified rotation period. This behaviour can be turned off by deselecting "Auto rotate planet". The rotation period can be changed at the bottom of this set of options with "Rotation Period", by either typing in a new number, or dragging the blue bar. By default one second of real time represents 3600 seconds (1 hour) of simulation time, so that the planet rotates once every 24 seconds. This can be changed through the "Speed" setting. Changing to smaller values with slow the simulation down towards real-time, increasing will speed up the simulation.  The seasons can also be automatically advanced with the "auto season" button and where the speed is controlled with the "Season Speed" slider.

Obliquity and Orbital Phase define the orientation of the planetary system. Obliquity changes the axial tilt of the planet, relative to the orbital plane, and orbital phase changes the position of the planet in it's orbit around the Sun. The dipole magnetic field can be adjusted with the five parameters in the main panel; they set the colatitude of the magnetic pole ("Dipole Colatitude"), the longitude of the magnetic pole ("Dipole Longitude") and the centre of the dipole ("Dipole Centre"). For example, Saturn's dipole has an offset of 0.04 planetary radii to the north, hence the centre would be 0.0 in the `x0` and `y0` widgets, and `0.04` in the `z0` widget.

Whilst these values can be adjusted in the GUI, specific planets and times can be set in the `Planets` panel. Click to open this up. Select a planet in the combobox and the configuration will automatically update to the current time. New widgets will also appear in the GUI to allow the user to show/hide different satellites/rings. The time can be changed in the `Time` widget; the time must be in the format YYYY-MM-DD. For each planet, a range of settings are also included under "Events". Selecting an event will automatically set up this scene. For example, the configuration of Neptune during the Voyager 2 flyby can be selected:

![alt text](neptune.png "Viewing Neptune at the time of writing; the view has been set to oblique using the 'Views' panel")

### Definitions
The unique orientation of Uranus can cause some difficulty in how the poles and rotation axis are defined. `Ikuchi` uses the IAU definition of a planet's north pole as the pole which lies is the same celestial hemisphere as Earth's north pole [e.g., [Archinal, B.A., et al., (2011) Celestial Mechanics and Dynamical Astronomy, 109(2), pp.101-135](https://link.springer.com/article/10.1007%2Fs10569-010-9320-4), and [Wikipedia](https://en.wikipedia.org/wiki/Poles_of_astronomical_bodies)]. According to this definition Uranus is a retrograde rotator and during the Voyager 2 encounter the south pole and the rotational pole were almost facing the Sun.

![alt text](uranus.png "Viewing Uranus at the Voyager 2 flyby")

### Customisation
The GUI controls include a number of customisation controls that allow the user to modify the colour scheme, opacity, the look of the orbital plane, and to enable and customise views of the equatorial and dipole equatorial planes.

## Rebuilding
`Ikuchi` requires no specific build process out of the box, but if you wish to change the time-period that `Ikuchi` has been setup for, or if you wish to use alternative textures, then you will need to use the `build.py` script in `source/build/`. This will build the 'ikuchi-pregenerated.js' file and will place it in `source/js/`. The build script requires `numpy`, `scipy`, `spiceypy` (https://github.com/AndrewAnnex/SpiceyPy), and optionally `matplotlib`. It also requires the appropriate textures from [Planet Pixel Emporium](http://planetpixelemporium.com/planets.html). The user might wish to use higher resolution textures; the supplied build uses the 1k textures from Planet Pixel Emporium.

## Future work
Features that are being developed are:
* Modify planets so that they are biaxial ellipsoids to account for the polar flattening of the giant planets.
* Colour surface by field strength.
* Add magnetic equator to planet surface.
* Add magnetopause and bow shock models.
* Add custom magnetic field models (e.g., traces from 'geopack' and spherical harmonic internal field models).
* Add current sheet surfaces.
* Add aurora rendering.
* Add time series widget to show dipole-Sun vector attack angle.
* Add aberrated solar wind angle vector.
* Compute geomagnetic field from IGRF with time.
* Add more accurate natural satellite orbits.
* Add option to switch between IAU and non-IAU definitions of planetary poles.
* Add Mercury, Venus and Mars.


CHANGES.md
# Change history

## Version 0.8 - 2023-05-15 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Added ability to automatically advance through seasons.
* Added an easily readable date element for use in generating movies, but which can be hidden.

## Version 0.7 - 2020-03-05 - Josh Wiggs <j.wiggs@lancaster.ac.uk> and Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Configuration data for each of the planets has been changed from a Javascript
file dumping data into the global space into a JSON file from which each
planet is setup dynamically.
* Meta data added to each planet including notes about origins of definitions.
* Added 'Show Information' button in the GUI to show meta data for each planet.
* CSS Fixes.
* Updated Uranus magnetic field definition to reflect that Uranus is rendered using
the IAU pole definition.  This was incorrectly applied in the previous version.
* Fixed bug where the NumberControllerBox in DAT.gui could not be adjusted using
the text boxes.  Added work-around by removing .listen() and adding an explicit
update when each planet changes (e.g., https://github.com/dataarts/dat.gui/issues/179).

## Version 0.6b - 2019-09-19 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Fixed moon orbit bug where moons would continue to orbit the planet when auto
rotate was turned off. Also fixed bug where moons would not respond to changes in speedup factor.
* Added GNU General Public Licence v3 to LICENCE.md and to each source file.
* Added favicon file.
* Moved library (e.g., THREE.js) javascript files into /js/lib folder from main javascript folder.
* Updated documentation in space physics NAIF frame kernel.
* Removed textures from distribution and adjusted build script accordingly.
* Removed ring optical depth to opacity functions from build script (not needed anymore).
* Added default planet rotation functions to build script.
* Extended meta tags in HTML.
* Updated source code documentation.
* Added modal about box and associated artwork.
* Added GPL v3 licence information.

## Version 0.5b - 2019-09-12 - Josh Wiggs <j.wiggs@lancaster.ac.uk>/Lancaster University
* Fixed field lines visibility switch
* Fixed moon orbit traces and field lines colour selectors
* Minor bug fixes.

## Version 0.5a - 2019-08-23 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Refactored code.
* Added moon rendering.
* Fixed memory leaks.
* Fixed orthographic camera bug.
* Modified how orbital and equatorial planes are rendered.
* Fixed ring rendering double-sided bug.
* Added dipole equatorial plane rendering.
* Removed optical depth->opacity ring rendering - just solid with a settable opacity.
* Changed discrete ring rendering to use lines rather than meshes.

## Version 0.4 - 2019-08-20 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Fixed scroll bar error.
* Added options to modify colours of different items.
* Added planetary ring rendering.
* Added orthographic camera and option to switch between orthographic and perspective views.
* Added build step which constructs a function for translating between ring optical depth and rendering opacity.

## Version 0.31 - 2019-08-16 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Fixed labelling error in Uranus solstices.
* Fixed GUI update error.

## Version 0.3 - 2019-08-16 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Added code to capture SVG version of onscreen image. Opens a new browser
	window with a textarea containing the SVG code that can be cut and pasted
	into a text file and then loaded into Illustrator or Inkscape.
* Clarified distinction between the north and south pole (according to the
	IAU definition; NB the north pole is the pole that lies in the same
	hemisphere off the ecliptic as Earth's north pole) and the rotational pole
	of the planet. Added functionality to reverse the rotation rate (e.g.,
	Uranus counter-rotates relative to the pole).
* Added sprites to highlight the north and south planetographic poles.

## Version 0.2 - 2019-08-12 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University
* Added code to visualise specific planets and specific times, with texture
	mapping and accurate orientations and rotation rates.
* Changed planetary rotation to a real-time model with a speedup factor.
* Added oblique viewing geometry option.
* Added event timings.
* Changed orbital plane rendering to a wireframe, rather than a shaded surface.
* Fixed equatorial plane flickering bug for zero obliquities.
* Added dipole offset as well as dipole orientation.

## Version 0.1 - 2019-03-23 - Chris Arridge <c.arridge@lancaster.ac.uk>/Lancaster University


source/js/ikuchi-fits.js
// Fits for Earth
function rotzEarth(t) {
	return(28969.76593826826 + -1.140795533078171e-08*t + 1.91455041474075*Math.sin(t*1.991063797294792e-10 + 166.45721759460946) + 0.020095207100803526*Math.sin(2*t*1.991063797294792e-10 + 3.0486487224410945));
}

function rotyEarth(t) {
	return(23.443179439452567 + -4.121869290538975e-15*t + -1.2272599199877887e-06*Math.sin(t*1.991063797294792e-10 + 0.9999820053054922));
}


// Fits for Neptune
function rotzNeptune(t) {
	return(255.91317766790257 + -6.893801246322523e-11*t + 0.9454730107723246*Math.sin(t*1.2083688173611776e-12 + -62.578127104062375) + -0.014744554948462245*Math.sin(2*t*1.2083688173611776e-12 + 1.266248446716515));
}

function rotyNeptune(t) {
	return(28.23499732092514 + 5.2904723464592514e-14*t + 0.034394668638443704*Math.sin(t*1.2083688173611776e-12 + 0.331170136656262));
}


// Fits for Jupiter
function rotzJupiter(t) {
	return(2363.210322546628 + -9.616143895679547e-10*t + -5.542986488709191*Math.sin(t*1.678489129283648e-11 + -411.3879611880666) + -0.16775457350349135*Math.sin(2*t*1.678489129283648e-11 + 12.882211363180492));
}

function rotyJupiter(t) {
	return(3.1190272601328144 + -4.084193371447696e-17*t + 5.860895728903148e-05*Math.sin(t*1.678489129283648e-11 + 0.6386993627562367));
}


// Fits for Uranus
function rotzUranus(t) {
	return(252.84767231373795 + -1.3577519209180985e-10*t + 5.413011491991998*Math.sin(t*2.369684154208593e-12 + 374.0894421221592) + -0.1524982565595177*Math.sin(2*t*2.369684154208593e-12 + 0.5073946730854844));
}

function rotyUranus(t) {
	return(82.22984393612073 + 7.291957245865771e-17*t + 0.000170581027977027*Math.sin(t*2.369684154208593e-12 + 0.537899776703132));
}


// Fits for Saturn
function rotzSaturn(t) {
	return(940.4269059601968 + -3.8743167223069975e-10*t + -6.207983106486679*Math.sin(t*6.759045002001111e-12 + 596.0437653880368) + 0.21686127313758174*Math.sin(2*t*6.759045002001111e-12 + 13.964708082127727));
}

function rotySaturn(t) {
	return(26.726932292193407 + 3.0378524217934277e-15*t + -0.0007065705139857624*Math.sin(t*6.759045002001111e-12 + 1.3102091218124712));
}


source/js/ikuchi-gui.js
/**
***	@file Provides all the GUI functions for Ikuchi.
***	@author Chris Arridge, Lancaster University <c.arridge@lancaster.ac.uk>
***	@version 0.31
***	@copyright Lancaster University (2019)
***	@licence TBD.
**/

// GUI global variables
var gui;
var guiFolderPlanets;
var guiFolderSetup;
var guiPlanetEvents;
var guiPlanetSatellites;

// Parameters that control the simulation, some of which are editable in the GUI.
var params = {
	// what L-shells do we show, how many field lines are there around
	// the planet, and do we show the field lines?
	fieldLshells: new Array(2.0,4.0,6.0,8.0),
	fieldNumLongitudes: 6,
	displayFieldLines: true,

	// should the planet auto-rotate and what is the speedup factor from
	// reality (3600 means in one second of real time, an hour has passed
	// in the simulated view).
	rotatePlanet: true,
	speedUp: 3600,

	// Colatitude and longitude of the dipole, and the origin of the dipole.
	// The string representation is for the GUI.
	fieldPoleColatitude: 10.0,
	fieldPoleLongitude: 0.0,
	fieldOriginX: 0.0,
	fieldOriginY: 0.0,
	fieldOriginZ: 0.0,
	fieldOrigin: '(0.0,0.0,0.0)',

	// Rotational state of the plaentary system (controlled by obliquity and
	// planetary phase) and the rotational phase and rotation period
	// of the planet.
	rotationY: 0.0,
	rotationZ: 0.0,
	rotationPhase: 0.0,
	rotationPeriod: 24.0,
	rotationReversed: false,

	// For the planet selection.
	planet: '- -',
	time: '- -',
	events: '- -',
	satellites: '- -',

	// Functions to place the view at a given position.
	setViewSun: function() {gCamera.position.set(10.0,0.0,0.0);},
	setViewDusk: function() {gCamera.position.set(0.0,10.0,0.0);},
	setViewDawn: function() {gCamera.position.set(0.0,-10.0,0.0);},
	setViewTop: function() {gCamera.position.set(0.0,0.0,10.0);},
	setViewOblque: function() {gCamera.position.set(5.77,5.77,5.77);},

	// Button to render to SVG. Inspired by https://www.marciot.com/blog-demos/three-to-svg/.
	goCaptureSVG: function () {
		var w = window.innerWidth;
		var h = window.innerHeight;
		var newWindow = window.open('', '', 'width='+w.toString()+', height='+h.toString());
		var svgCodeContainer = newWindow.document.createElement('textarea');
		svgCodeContainer.id = 'source';
		svgCodeContainer.cols = 120;
		svgCodeContainer.rows = 40;
		newWindow.document.body.appendChild(svgCodeContainer)

		svgRenderer = new THREE.SVGRenderer();
		svgRenderer.setClearColor(0xffffff);
		svgRenderer.setSize(w, h);
		svgRenderer.setQuality('low');
		svgCodeContainer.appendChild(svgRenderer.domElement);
		svgRenderer.render(gScene,gCamera);

		newWindow.document.getElementById('source').value = svgCodeContainer.innerHTML.replace(/<path/g,"\n<path");
	},

	// Move the simulation to a given planet at a given time.
	goPlanet: function() {
		// Convert the time into a unix timestamp (in milliseconds) so that
		// the fits can be used to figure out the obliquity and orbital phase
		// angles.
		t = new Date(params.time);

		// Giant planet dipole position and orientation taken from
		// Connerney (1993) JGR 98(E10), 18659-18679.
		switch(params.planet) {
			case 'Earth':
				params.rotationPeriod = 24.0;
				updateRotationReversed(false);
				params.rotationY = rotyEarth(t.getTime());
				params.rotationZ = rotzEarth(t.getTime()) % 360.0;
				updateFieldColat(10.1);
				updateFieldLong(288.0);
				updateFieldOrigin('(0.0,0.0,0.0)');
				gPlanetImageElement.src = textureEarthDayMap;
				break;
			case 'Jupiter':
				params.rotationPeriod = 9.925;
				updateRotationReversed(false);
				params.rotationY = rotyJupiter(t.getTime());
				params.rotationZ = rotzJupiter(t.getTime()) % 360.0;
				updateFieldOrigin('(-0.01,0.0,-0.01)');
				updateFieldColat(10.8);
				updateFieldLong(201.0);
				gPlanetImageElement.src = textureJupiter;
				break;
			case 'Saturn':
				params.rotationPeriod = 10.0 + (33.0*60 + 38.0)/3600.0;
				updateRotationReversed(false);
				params.rotationY = rotySaturn(t.getTime());
				params.rotationZ = rotzSaturn(t.getTime()) % 360.0;
				updateFieldOrigin('(0.0,0.0,0.04)');
				updateFieldColat(0.0);
				updateFieldLong(0.0);
				gPlanetImageElement.src = textureSaturn;
				break;
			case 'Uranus':
				params.rotationPeriod = 17.0 + (14.0*60 + 24.0)/3600.0;
				updateRotationReversed(true);
				params.rotationY = rotyUranus(t.getTime());
				params.rotationZ = rotzUranus(t.getTime()) % 360.0;
				updateFieldOrigin('(-0.02,0.02,-0.31)');
				updateFieldColat(60.0);
				updateFieldLong(48.0);		// fixme
				gPlanetImageElement.src = textureUranus;
				break;
			case 'Neptune':
				params.rotationPeriod = 16.0 + (6.0*60 + 36.0)/3600.0;
				updateRotationReversed(false);
				params.rotationY = rotyNeptune(t.getTime());
				params.rotationZ = rotzNeptune(t.getTime()) % 360.0;
				updateFieldOrigin('(0.17,0.46,-0.24)');
				updateFieldColat(46.8);
				updateFieldLong(79.5);
				gPlanetImageElement.src = textureNeptune;
				break;
		}
	}
};


/**
*** Initialise the GUI.
***
*** @todo: add capability to select rings/satellites in the planet selector.
*** @todo: add capability to change the colours.
**/
function initGui()
{
	gui = new dat.GUI({height : 5 * 32 - 1});

	// main control parmeters
	gui.add(params, 'rotatePlanet').name('Auto rotate planet')
	gui.add(params, 'speedUp').name('Speed').min(100.0).max(10000).step(1.0).listen();
	gui.add(params, 'rotationY').name('Obliquity').min(-90.0).max(90.0).step(5.0).onChange(updateObliquity).listen();
	gui.add(params, 'rotationZ').name('Orbital Phase').min(0.0).max(360.0).step(10.0).onChange(updateOrbitPhase).listen();
	gui.add(params, 'fieldPoleColatitude').name('Dipole Colatitude').min(0.0).max(180.0).step(5.0).onChange(updateFieldColat).listen();
	gui.add(params, 'fieldPoleLongitude').name('Dipole Longitude').min(0.0).max(360.0).step(5.0).onChange(updateFieldLong).listen();
	gui.add(params, 'fieldOrigin').name('Dipole Centre').onFinishChange(updateFieldOrigin).listen();
	gui.add(params, 'rotationPeriod').name('Rotation Period').min(9.0).max(29.0).step(0.05).listen();
	gui.add(params, 'rotationReversed').name('Reverse Rotation').onChange(updateRotationReversed).listen();
	gui.add(params, 'goCaptureSVG').name('Capture SVG Code');

	// quick button to change views
	var guiViewFolders = gui.addFolder('Views');
	guiViewFolders.add(params, 'setViewSun').name('Subsolar point');
	guiViewFolders.add(params, 'setViewDusk').name('Dusk meridian');
	guiViewFolders.add(params, 'setViewDawn').name('Dawn meridian');
	guiViewFolders.add(params, 'setViewTop').name('Overhead');
	guiViewFolders.add(params, 'setViewOblque').name('Oblique');

	// setup presets
	guiFolderSetup = gui.addFolder('Setup');
	guiFolderSetup.add(params, 'fieldNumLongitudes').name('Num. longitudes')

	// select specific planets
	guiFolderPlanets = gui.addFolder('Planets');
	guiFolderPlanets.add(params, 'planet', ['- -','Earth','Jupiter','Saturn','Uranus','Neptune']).name('Planet').onFinishChange(updatePlanet);
	guiFolderPlanets.add(params, 'time').name('Time').listen();
	guiFolderPlanets.add(params, 'goPlanet').name('Go');
	guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['- -']).name('Events').onFinishChange(updatePlanetEvent);
//	guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['- -']).name('Satellites').onFinishChange(updatePlanetSatellites)
}


/**
*** Respond to a planet having been selected in the planet combobox. This sets
*** the current time in the time box, adds specific events to the event
*** combobox.
***
*** @todo: add capability to select rings/satellites in the planet selector.
**/
function updatePlanet(val) {
	var now = new Date();
	var nowStr = now.getFullYear() + '-' + String(now.getMonth()+1).padStart(2,'0') + '-' + String(now.getDate()).padStart(2,'0')

	switch(val) {
		case '- -':
			params.time = '';
//			guiPlanetEvents.remove()
//			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['- -']).name('Events');
			break;
		case 'Earth':
			params.time = nowStr;
			guiPlanetEvents.remove()
			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['Now','Vernal Equinox','NH Summer Solstice','Autumnal Equinox','SH Summer Solstice']).name('Events').onFinishChange(updatePlanetEvent);
//			guiPlanetSatellites.remove()
//			guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['None','Moon']).name('Satellites')
			break;
		case 'Jupiter':
			params.time = nowStr;
			guiPlanetEvents.remove()
			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['Now']).name('Events').onFinishChange(updatePlanetEvent);
//			guiPlanetSatellites.remove()
//			guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['None','Io/Eur/Gan/Cal']).name('Satellites')
			break;
		case 'Saturn':
			params.time = nowStr;
			guiPlanetEvents.remove()
			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['Now','Vernal Equinox','NH Summer Solstice','Autumnal Equinox','SH Summer Solstice','Cassini SOI']).name('Events').onFinishChange(updatePlanetEvent);
//			guiPlanetSatellites.remove()
//			guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['None','Enc/Tit','Mim/Enc/Tet/Dio/Rhe/Tit/Iap']).name('Satellites')
			break;
		case 'Uranus':
			params.time = nowStr;
			guiPlanetEvents.remove()
			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['Now','Voyager 2','Vernal Equinox','NH Summer Solstice','Autumnal Equinox','SH Summer Solstice',]).name('Events').onFinishChange(updatePlanetEvent);
//			guiPlanetSatellites.remove()
//			guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['None','Mir/Ari/Umb/Tit/Obe']).name('Satellites')
			break;
		case 'Neptune':
			params.time = nowStr;
			guiPlanetEvents.remove()
			guiPlanetEvents = guiFolderPlanets.add(params, 'events', ['Now','Voyager 2','Vernal Equinox','NH Summer Solstice','Autumnal Equinox','SH Summer Solstice',]).name('Events').onFinishChange(updatePlanetEvent);
//			guiPlanetSatellites.remove()
//			guiPlanetSatellites = guiFolderPlanets.add(params, 'satellites', ['None','Triton','Pro/Tri/Ner']).name('Satellites')
			break;
		default:
			throw new Error("Unknown planet name");
	}
}



/**
*** Called when the rotation orientation of the planet is changed. Sets the
*** parameter and changes the rotation axis vector object.
**/
function updateRotationReversed(val) {
	params.rotationReversed = val;
	if (params.rotationReversed) {
		gScene.getObjectByName('Rotation Axis Vector').rotation.set(-Math.PI/2.0,0.0,0.0);
	} else {
		gScene.getObjectByName('Rotation Axis Vector').rotation.set(Math.PI/2.0,0.0,0.0);
	}
	gScene.getObjectByName('Rotation Axis Vector').updateMatrix();
}


/**
*** Called when the obliquity of the planet is directly changed in the
*** GUI. Sets the parameter and adjusted the rotation matrices.
**/
function updateObliquity(val) {
	params.rotationY = val;
	setPlanetSystemRotationMatrix();
}


/**
*** Called when the orbital phase of the planet is directly changed in the
*** GUI. Sets the parameter and adjusted the rotation matrices.
**/
function updateOrbitPhase(val) {
	params.rotationZ=val;
	setPlanetSystemRotationMatrix();
}


/**
*** Called when the pole colatitude of the dipole is directly changed in the
*** GUI. Sets the parameter and adjusted the rotation matrices.
**/
function updateFieldColat(val) {
	params.fieldPoleColatitude = val;
	var yMatrix = new THREE.Matrix4().makeRotationY(params.fieldPoleColatitude*Math.PI/180.0);
	var zMatrix = new THREE.Matrix4().makeRotationZ(params.fieldPoleLongitude*Math.PI/180.0);
	var transMatrix = new THREE.Matrix4().makeTranslation(params.fieldOriginX,params.fieldOriginY,params.fieldOriginZ);
	gScene.getObjectByName('Dipole plus Vector').matrix = transMatrix.multiply(zMatrix.multiply(yMatrix));
	setPlanetSystemRotationMatrix();
}


/**
*** Called when the pole longitude of the dipole is directly changed in the
*** GUI. Sets the parameter and adjusted the rotation matrices.
**/
function updateFieldLong(val) {
	params.fieldPoleLongitude = val;
	var yMatrix = new THREE.Matrix4().makeRotationY(params.fieldPoleColatitude*Math.PI/180.0);
	var zMatrix = new THREE.Matrix4().makeRotationZ(params.fieldPoleLongitude*Math.PI/180.0);
	var transMatrix = new THREE.Matrix4().makeTranslation(params.fieldOriginX,params.fieldOriginY,params.fieldOriginZ);
	gScene.getObjectByName('Dipole plus Vector').matrix = transMatrix.multiply(zMatrix.multiply(yMatrix));
	setPlanetSystemRotationMatrix();
}


/**
*** Called when the origin of the dipole is directly changed in the
*** GUI. Sets the parameter and adjusted the rotation matrices.
**/
function updateFieldOrigin(val) {
	params.fieldOrigin = val;
	v = val.replace(/[()\s]+/g, '').split(',')
	params.fieldOriginX = Number(v[0]);
	params.fieldOriginY = Number(v[1]);
	params.fieldOriginZ = Number(v[2]);
	var yMatrix = new THREE.Matrix4().makeRotationY(params.fieldPoleColatitude*Math.PI/180.0);
	var zMatrix = new THREE.Matrix4().makeRotationZ(params.fieldPoleLongitude*Math.PI/180.0);
	var transMatrix = new THREE.Matrix4().makeTranslation(params.fieldOriginX,params.fieldOriginY,params.fieldOriginZ);
	gScene.getObjectByName('Dipole plus Vector').matrix = transMatrix.multiply(zMatrix.multiply(yMatrix));
	setPlanetSystemRotationMatrix();
}


/**
*** Called when the a planet "event" is selected in the GUI combobox. Sets the
*** appropriate time (e.g., closest approach time of a spacecraft) and then
*** invokes goPlanet to set everything up.
**/
function updatePlanetEvent(val) {
	var now = new Date();
	var nowStr = now.getFullYear() + '-' + String(now.getMonth()+1).padStart(2,'0') + '-' + String(now.getDate()).padStart(2,'0')
	console.log(val);
	switch(params.planet) {
		// Equinox and solstice times from https://en.wikipedia.org/wiki/Equinox
		case 'Earth':
			switch(val) {
				case 'Now':
					params.time = nowStr;
					params.goPlanet();
					break;
				case 'NH Summer Solstice':
					params.time = '2019-06-21';
					params.goPlanet();
					break;
				case 'SH Summer Solstice':
					params.time = '2019-12-22';
					params.goPlanet();
					break;
				case 'Vernal Equinox':
					params.time = '2019-03-20';
					params.goPlanet();
					break;
				case 'Autumnal Equinox':
					params.time = '2019-09-23';
					params.goPlanet();
					break;
			}
			break;

		case 'Jupiter':
			switch(val) {
				case 'Now':
					params.time = nowStr;
					params.goPlanet();
					break;
			}
			break;

		case 'Saturn':
			switch(val) {
				case 'Now':
					params.time = nowStr;
					params.goPlanet();
					break;
				case 'Vernal Equinox':
					params.time = '2009-08-11';
					params.goPlanet();
					break;
				case 'NH Summer Solstice':
					params.time = '2018-05-23';
					params.goPlanet();
					break;
				case 'Autumnal Equinox':
					params.time = '1995-11-19';
					params.goPlanet();
					break;
				case 'SH Summer Solstice':
					params.time = '2002-10-25';
					params.goPlanet();
					break;
				case 'Cassini SOI':
					params.time = '2004-07-01';
					params.goPlanet();
					break;
			}
			break;

		// Equinox times from: Meeus, J (1997) 'Equinoxes and solstices on
		// Uranus and Neptune' J. Brit. Astronomical Assoc. 107(6), p332,
		// http://adsabs.harvard.edu/full/1997JBAA..107..332M
		case 'Uranus':
			switch(val) {
				case 'Now':
					params.time = nowStr;
					params.goPlanet();
					break;
				case 'Autumnal Equinox':
					params.time = '1966-02-03';
					params.goPlanet();
					break;
				case 'SH Summer Solstice':
					params.time = '1985-10-06';
					params.goPlanet();
					break;
				case 'Vernal Equinox':
					params.time = '2007-12-16';
					params.goPlanet();
					break;
				case 'NH Summer Solstice':
					params.time = '2030-04-19';
					params.goPlanet();
					break;
				case 'Voyager 2':
					params.time = '1986-01-24';
					params.goPlanet();
					break;
			}
			break;

		// Equinox times from: Meeus, J (1997) 'Equinoxes and solstices on
		// Uranus and Neptune' J. Brit. Astronomical Assoc. 107(6), p332,
		// http://adsabs.harvard.edu/full/1997JBAA..107..332M
		case 'Neptune':
			switch(val) {
				case 'Now':
					params.time = nowStr;
					params.goPlanet();
					break;
				case 'Vernal Equinox':
					params.time = '2038-02-28';
					params.goPlanet();
					break;
				case 'NH Summer Solstice':
					params.time = '2078-11-14';
					params.goPlanet();
					break;
				case 'Autumnal Equinox':
					params.time = '1955-04-11';
					params.goPlanet();
					break;
				case 'SH Summer Solstice':
					params.time = '1997-02-26';
					params.goPlanet();
					break;
				case 'Voyager 2':
					params.time = '1989-08-25';
					params.goPlanet();
					break;
			}
			break;
	}
}
